src/vector/sources/disks/disk_queue_length.rs

```diff
@@ -192,3 +182,113 @@ impl SourceConfig for DiskQueueLengthConfig {
+        struct Test {
+            pub diskstats: &'static str,
+            pub expected_results: Vec<DiskQueueLengthResult>,
+        }
+
+        let tests = vec![
+            // A typical diskstats file
+            Test {
+                diskstats: r#"
+   1       0 loop0 0 0 0  0  0 0  0  0   0 0 0 0 0 0 0
+   1       1 loop1 0 0 0  0  0 0  0  0   0 0 0 0 0 0 0
+   1       2 loop2 0 0 0  0  0 0  0  0   0 0 0 0 0 0 0   
+   2       0 sda   1 0 4  8  9 0 13 14   2 1 2 0 0 0 0
+   3       2 sda1  2 0 5 10 12 0 15  1   1 2 2 0 0 0 0
+   4       3 sda2  3 0 6  7  0 0  0  0   0 1 2 0 0 0 0
+     "#,
+                expected_results: vec![
+                    DiskQueueLengthResult {
+                        disk: "sda".to_string(),
+                        value: 2.0,
+                    },
+                    DiskQueueLengthResult {
+                        disk: "sda1".to_string(),
+                        value: 1.0,
+                    },
+                    DiskQueueLengthResult {
+                        disk: "sda2".to_string(),
+                        value: 0.0,
+                    },
+                ],
+            },
+            Test {
+                diskstats: r#"
+     "#,
+                expected_results: vec![],
+            },
+            // This one has a float instead of an integer for sda2
+            Test {
+                diskstats: r#"
+   1       0 loop0 0 0 0  0  0 0  0  0   0 0 0 0 0 0 0
+   1       1 loop1 0 0 0  0  0 0  0  0   0 0 0 0 0 0 0
+   1       2 loop2 0 0 0  0  0 0  0  0   0 0 0 0 0 0 0   
+   2       0 sda   1 0 4  8  9 0 13 14   2 1 2 0 0 0 0
+   3       2 sda1  2 0 5 10 12 0 15  1   1 2 2 0 0 0 0
+   4       3 sda2  3 0 6  7  0 0  0  0 4.5 1 2 0 0 0 0
+     "#,
+                expected_results: vec![
+                    DiskQueueLengthResult {
+                        disk: "sda".to_string(),
+                        value: 2.0,
+                    },
+                    DiskQueueLengthResult {
+                        disk: "sda1".to_string(),
+                        value: 1.0,
+                    },
+                ],
+            },
+            // this one is malformed
+            Test {
+                diskstats: r#"
+   1       0 loop0 0 0 0  0  0 0  0  0   0 0 0 0 0 0 0
+   1       1 loop1 0 0 0  0  0 0  0  0   0 0 0 0 0 0 0
+   1       2 loop2 0 0 0  0  0 0  0  0   0 0 0 0 0 0 0   
+   2       0 sda   1 0 4  8 
+   3       2 sda1  2 0 5 10 12 0 15  1   1 2 2 0 0 0 0
+   4  
+   5       3 sda2  
+     "#,
+                expected_results: vec![DiskQueueLengthResult {
+                    disk: "sda1".to_string(),
+                    value: 1.0,
+                }],
+            },
+        ];
+
+        for test in tests {
+            let mut file = tempfile::NamedTempFile::new().expect("couldn't make temp file");
+            let r = std::io::Write::write(&mut file, test.diskstats.as_bytes());
+            std::io::Write::flush(&mut file).expect("flush failed");
```

Comment:
Is there a reason you have to use a file here?

Reply:
Since creating a temporary file in the tests was trivial I opted to do that and make the file path get sent as an argument to `get_disk_queue_length`, which kept me from having to change the function too much as I extracted it. I suppose an alternative idea would be to have changed `get_disk_queue_length` to accept a buffer or string or something and read the file at the call site, but that seemed like more work than needed.